src/core/reactor_backend.cc

```diff
@@ -1478,51 +1493,149 @@ class reactor_backend_uring final : public reactor_backend {
+                        fd.speculate_epoll(EPOLLIN);
+                    }
+                    return make_ready_future<size_t>(*r);
+                }
+            } catch (const std::system_error& e) {
+                return make_exception_future<size_t>(e);
             }
-            if (size_t(*r) == internal::iovec_len(iov)) {
-                fd.speculate_epoll(EPOLLIN);
+        }
+        class read_completion final : public io_completion {
+            pollable_fd_state& _fd;
+            std::vector<iovec> _iov;
+            ::msghdr _mh = {};
+            promise<size_t> _result;
+        public:
+            read_completion(pollable_fd_state& fd, const std::vector<iovec>& iov)
+                : _fd(fd), _iov(iov) {
+                _mh.msg_iov = const_cast<iovec*>(_iov.data());
+                _mh.msg_iovlen = _iov.size();
             }
-            return make_ready_future<size_t>(*r);
-        });
+            void complete(size_t bytes) noexcept final {
+                if (bytes == internal::iovec_len(_iov)) {
+                    _fd.speculate_epoll(EPOLLIN);
+                }
+                _result.set_value(bytes);
+                delete this;
+            }
+            void set_exception(std::exception_ptr eptr) noexcept final {
+                _result.set_exception(eptr);
+                delete this;
+            }
+            ::msghdr* msghdr() {
+                return &_mh;
+            }
+            future<size_t> get_future() {
+                return _result.get_future();
+            }
+        };
+        auto desc = std::make_unique<read_completion>(fd, iov);
+        auto req = internal::io_request::make_recvmsg(fd.fd.get(), desc->msghdr(), 0);
+        return submit_request(std::move(desc), std::move(req));
     }
     virtual future<temporary_buffer<char>> read_some(pollable_fd_state& fd, internal::buffer_allocator* ba) override {
-        return readable(fd).then([this, &fd, ba] {
+        if (fd.take_speculation(POLLIN)) {
             auto buffer = ba->allocate_buffer();
-            auto r = fd.fd.read(buffer.get_write(), buffer.size());
-            if (!r) {
-                // Speculation failure, try again with real polling this time
-                // Note we release the buffer and will reallocate it when poll
-                // completes.
-                return read_some(fd, ba);
-            }
-            if (size_t(*r) == buffer.size()) {
-                fd.speculate_epoll(EPOLLIN);
+            try {
+                auto r = fd.fd.read(buffer.get_write(), buffer.size());
+                if (r) {
+                    if (size_t(*r) == buffer.size()) {
+                        fd.speculate_epoll(EPOLLIN);
+                    }
+                    return make_ready_future<temporary_buffer<char>>(std::move(buffer));
+                }
+            } catch (const std::system_error& e) {
+                return make_exception_future<temporary_buffer<char>>(e);
```

Comment:
You can use `catch (...)` and `current_exception_as_future()`, they're safer. In particular copying an exception caught by reference can lead to object slicing.

Reply:
okay. wanted to make it explicit by capturing only the expected exception type. will change in the next revision.